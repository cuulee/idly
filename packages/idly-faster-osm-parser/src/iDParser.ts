import { lngLatFactory } from 'idly-common/lib/geo/lngLatFactory';
import { nodeFactory } from 'idly-common/lib/osm/entityFactory/nodeFactory';
import { relationFactory } from 'idly-common/lib/osm/entityFactory/relationFactory';
import { wayFactory } from 'idly-common/lib/osm/entityFactory/wayFactory';

import { relationMemberFactory } from 'idly-common/lib/osm/relationMemberFactory';
import {
  Entity,
  LngLat,
  Node as OSMNode,
  Relation,
  RelationMember,
  Way,
} from 'idly-common/lib/osm/structures';

/**
 * Converts the osm xml to an array of Entity,
 * This implementation is a port from iD.
 */
export function iDParser(xml: Document): Entity[] {
  if (!xml || !xml.childNodes) {
    return [];
  }
  let root = xml.childNodes[2];
  if (!root) {
    root = xml.childNodes[0];
  }
  const children = root.childNodes;
  const entities: Entity[] = [];
  const len = children.length;
  for (let i = 0; i < len; i++) {
    const child = children[i];
    if (child.nodeName === 'node') {
      entities.push(nodeData(child));
    } else if (child.nodeName === 'relation') {
      entities.push(relationData(child));
    } else if (child.nodeName === 'way') {
      entities.push(wayData(child));
    }
  }
  return entities;
}
/**
 * @param attrs
 */
function getVisible(attrs: NamedNodeMap): boolean {
  return (
    !attrs.getNamedItem('visible') ||
    attrs.getNamedItem('visible').value !== 'false'
  );
}

function getMembers(obj: any): RelationMember[] {
  const elems = obj.getElementsByTagName('member');
  const members: RelationMember[] = [];
  const len = elems.length;
  for (let i = 0; i < len; i++) {
    const attrs = elems[i].attributes;
    members.push(
      relationMemberFactory({
        id:
          attrs.getNamedItem('type').value[0] + attrs.getNamedItem('ref').value,
        role: attrs.getNamedItem('role').value,
        type: attrs.getNamedItem('type').value,
      })
    );
  }
  return members;
}

function getNodes(obj: any): string[] {
  const elems = obj.getElementsByTagName('nd');
  const nodes: string[] = [];
  const len = elems.length;
  for (let i = 0; i < len; i++) {
    nodes.push('n' + elems[i].attributes.getNamedItem('ref').value);
  }
  return nodes;
}

function getTags(obj: any): { [index: string]: string } {
  const elems = obj.getElementsByTagName('tag');
  const t: any = {};

  for (let i = 0, l = elems.length; i < l; i++) {
    const attrs = elems[i].attributes;
    t[attrs.getNamedItem('k').value] = attrs.getNamedItem('v').value;
  }
  return t;
}

function getLoc(attrs: NamedNodeMap): LngLat {
  const lon = attrs.getNamedItem('lon') && attrs.getNamedItem('lon').value;
  const lat = attrs.getNamedItem('lat') && attrs.getNamedItem('lat').value;
  return lngLatFactory([parseFloat(lon), parseFloat(lat)]);
}

function nodeData(obj: Node): OSMNode {
  const attrs = obj.attributes;
  return nodeFactory(
    {
      attributes: {
        changeset: attrs.getNamedItem('changeset')
          ? attrs.getNamedItem('changeset').value
          : undefined,
        timestamp: attrs.getNamedItem('timestamp')
          ? attrs.getNamedItem('timestamp').value
          : undefined,
        uid: attrs.getNamedItem('uid')
          ? attrs.getNamedItem('uid').value
          : undefined,
        user: attrs.getNamedItem('user')
          ? attrs.getNamedItem('user').value
          : undefined,
        version: attrs.getNamedItem('version').value,
        visible: getVisible(attrs) as boolean,
      },
      id: 'n' + attrs.getNamedItem('id').value,
      loc: getLoc(attrs),
      tags: getTags(obj),
    },
    false
  );
}

function relationData(obj: Node): Relation {
  const attrs = obj.attributes;
  return relationFactory(
    {
      attributes: {
        changeset: attrs.getNamedItem('changeset')
          ? attrs.getNamedItem('changeset').value
          : undefined,
        timestamp: attrs.getNamedItem('timestamp')
          ? attrs.getNamedItem('timestamp').value
          : undefined,
        uid: attrs.getNamedItem('uid')
          ? attrs.getNamedItem('uid').value
          : undefined,
        user: attrs.getNamedItem('user')
          ? attrs.getNamedItem('user').value
          : undefined,
        version: attrs.getNamedItem('version').value,
        visible: getVisible(attrs),
      },
      id: 'r' + attrs.getNamedItem('id').value,
      members: getMembers(obj),
      tags: getTags(obj),
    },
    false
  );
}

function wayData(obj: Node): Way {
  const attrs = obj.attributes;
  return wayFactory(
    {
      attributes: {
        changeset: attrs.getNamedItem('changeset')
          ? attrs.getNamedItem('changeset').value
          : undefined,
        timestamp: attrs.getNamedItem('timestamp')
          ? attrs.getNamedItem('timestamp').value
          : undefined,
        uid: attrs.getNamedItem('uid')
          ? attrs.getNamedItem('uid').value
          : undefined,
        user: attrs.getNamedItem('user')
          ? attrs.getNamedItem('user').value
          : undefined,
        version: attrs.getNamedItem('version').value,
        visible: getVisible(attrs),
      },
      id: 'w' + attrs.getNamedItem('id').value,
      nodes: getNodes(obj),
      tags: getTags(obj),
    },
    false
  );
}
