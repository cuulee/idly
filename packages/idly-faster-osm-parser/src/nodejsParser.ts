import * as fs from 'fs';
import genReadlines = require('gen-readlines');

import { Entity } from 'idly-common/lib/osm/structures';

import { Options } from './helpers';
import { objToStandard, xmlLineToObj } from './process';

function* lineParser(path: string) {
  const fd = fs.openSync(path, 'r');
  const stats = fs.fstatSync(fd);
  const reader = genReadlines(fd, stats.size);
  for (const line of reader) {
    yield line.toString();
  }
  fs.closeSync(fd);
}

/**
 * The parser for nodejs. It directly reads a file
 * line by line. Reading a the file line by line enables
 * this to load any kind of huge file.
 */
export function* nodejsParser(
  path: string,
  opts: Options = {}
): IterableIterator<Entity> {
  opts = {
    prependEntityChar: true,
    useRef: false,
    freeze: false,
    ...opts,
  };

  const lastObj: any = {
    pointer: undefined,
  };

  for (const r of lineParser(path)) {
    const result = xmlLineToObj(r, lastObj, opts);
    if (result) {
      yield objToStandard(result, opts);
    }
  }
}
