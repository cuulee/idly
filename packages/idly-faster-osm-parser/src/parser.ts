import { Entity } from 'idly-common/lib/osm/structures';
import { Options } from './helpers';
import { objToStandard, xmlLineToObj } from './process';

export default parser;

/**
 * A an extremely fast smart parserÂ for .osm api
 */
export function parser(str: string, opts: Options = {}): Entity[] {
  const lastObj: any = {
    pointer: undefined,
  };

  opts = {
    prependEntityChar: true,
    useRef: false,
    freeze: false,
    ...opts,
  };

  return str
    .split('\n')
    .map(r => {
      const result = xmlLineToObj(r, lastObj, opts);
      if (result) {
        return result;
      }
    })
    .filter(r => r)
    .map(r => objToStandard(r, opts));
}
