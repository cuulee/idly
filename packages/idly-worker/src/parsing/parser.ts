import { deepFreeze } from 'idly-common/lib/misc/deepFreeze';
import { ImMap, ImSet } from 'idly-common/lib/misc/immutable';
import { attributesGen } from 'idly-common/lib/osm/attributesGen';
import { genLngLat } from 'idly-common/lib/osm/genLngLat';
import { nodeFactory } from 'idly-common/lib/osm/nodeFactory';
import { relationFactory } from 'idly-common/lib/osm/relationFactory';
import { relationMemberGen } from 'idly-common/lib/osm/relationMemberGen';
import {
  Entity,
  EntityId,
  EntityTable,
  Node,
  ParentWays,
  Relation,
  Tags,
  Way
} from 'idly-common/lib/osm/structures';
import { tagsFactory } from 'idly-common/lib/osm/tagsFactory';
import { wayFactory } from 'idly-common/lib/osm/wayFactory';

export function calculateParentWays(parentWays: ParentWays, ways: Way[]) {
  console.time('Imm: calculateParentWays');
  const x = parentWays.withMutations(p => {
    ways.forEach(w => {
      w.nodes.forEach(nodeId => {
        p.update(nodeId, (s = ImSet()) => s.add(w.id));
      });
    });
  });
  console.timeEnd('Imm: calculateParentWays');
  return x;
}

const xmlIndex = process.env.NODE_ENV === 'test' ? 0 : 2;
/**
 * @NOTE parseXML doesnt care if there are duplicates
 *  as this helps in removing a particular tile
 *  and still maintaining the duplicated entity in other
 *  tile. If in case we started screening duplicates
 *  this might remove an entity which might span across
 *  multiple tiles.
 * @param xml
 * @param parentWays
 */
export function parseXML(
  xml: Document | undefined,
  parentWays: ParentWays = ImMap()
):
  | undefined
  | {
      entities: Entity[];
      parentWays: ParentWays;
    } {
  if (!xml || !xml.childNodes) return;
  const root = xml.childNodes[xmlIndex];

  const children = root.childNodes;
  const entities: Entity[] = [];
  const nodesXML: any = [];
  const waysXML: any = [];
  const relationsXML: any = [];
  const group: {
    node: Node[];
    way: Way[];
    relation: Relation[];
  } = {
    node: [],
    way: [],
    relation: []
  };
  for (let i = 0, l = children.length; i < l; i++) {
    const child = children[i];
    const parser = group[child.nodeName];
    if (parser) group[child.nodeName].push(child);
  }

  group.relation = group.relation.map(parsers.relation);
  group.way = group.way.map(w => parsers.way(w));

  // calculateParentWays right at this point allows us
  // to pass the parsed group.way which is needed to
  // calculate parentWays.
  // @NOTE calculateParentWays directly mutates
  //  I want to change this behaviour
  // parentWays = calculateParentWays(parentWays, group.way);

  parentWays = calculateParentWays(parentWays, group.way);

  group.node = group.node.map(n => parsers.node(n));

  return {
    entities: [...group.node, ...group.way, ...group.relation],
    parentWays
  };
}

function getVisible(attrs) {
  return (
    !attrs.getNamedItem('visible') ||
    attrs.getNamedItem('visible').value !== 'false'
  );
}

function getMembers(obj) {
  const elems = obj.getElementsByTagName('member');
  const members = new Array(elems.length);
  for (let i = 0, l = elems.length; i < l; i++) {
    const attrs = elems[i].attributes;
    members[i] = relationMemberGen({
      id: attrs.getNamedItem('type').value[0] + attrs.getNamedItem('ref').value,
      type: attrs.getNamedItem('type').value,
      role: attrs.getNamedItem('role').value
    });
  }
  return members;
}

function getNodes(obj) {
  const elems = obj.getElementsByTagName('nd');
  const nodes = new Array(elems.length);
  for (let i = 0, l = elems.length; i < l; i++) {
    nodes[i] = 'n' + elems[i].attributes.getNamedItem('ref').value;
  }
  return nodes;
}

function getTags(obj): Tags {
  const elems = obj.getElementsByTagName('tag');
  const t: Array<[string, string]> = [];
  for (let i = 0, l = elems.length; i < l; i++) {
    const attrs = elems[i].attributes;
    t.push([attrs.getNamedItem('k').value, attrs.getNamedItem('v').value]);
  }
  return tagsFactory(t);
}

function getLoc(attrs) {
  const lon = attrs.getNamedItem('lon') && attrs.getNamedItem('lon').value;
  const lat = attrs.getNamedItem('lat') && attrs.getNamedItem('lat').value;
  return genLngLat([parseFloat(lon), parseFloat(lat)]);
}

const parsers = {
  node: function nodeData(obj) {
    const attrs = obj.attributes;
    const id = 'n' + attrs.getNamedItem('id').value;
    return nodeFactory({
      id,
      attributes: attributesGen({
        visible: getVisible(attrs),
        version: attrs.getNamedItem('version').value,
        timestamp:
          attrs.getNamedItem('timestamp') &&
          attrs.getNamedItem('timestamp').value,
        changeset:
          attrs.getNamedItem('changeset') &&
          attrs.getNamedItem('changeset').value,
        uid: attrs.getNamedItem('uid') && attrs.getNamedItem('uid').value,
        user: attrs.getNamedItem('user') && attrs.getNamedItem('user').value
      }),
      loc: getLoc(attrs),
      tags: getTags(obj)
    });
  },

  way: function wayData(obj) {
    const attrs = obj.attributes;
    return wayFactory({
      id: 'w' + attrs.getNamedItem('id').value,
      attributes: attributesGen({
        visible: getVisible(attrs),
        version: attrs.getNamedItem('version').value,
        changeset:
          attrs.getNamedItem('changeset') &&
          attrs.getNamedItem('changeset').value,
        timestamp:
          attrs.getNamedItem('timestamp') &&
          attrs.getNamedItem('timestamp').value,
        user: attrs.getNamedItem('user') && attrs.getNamedItem('user').value,
        uid: attrs.getNamedItem('uid') && attrs.getNamedItem('uid').value
      }),
      tags: getTags(obj),
      nodes: getNodes(obj)
    });
  },

  relation: function relationData(obj) {
    const attrs = obj.attributes;
    return relationFactory({
      id: 'r' + attrs.getNamedItem('id').value,
      attributes: attributesGen({
        visible: getVisible(attrs),
        version: attrs.getNamedItem('version').value,
        changeset:
          attrs.getNamedItem('changeset') &&
          attrs.getNamedItem('changeset').value,
        timestamp:
          attrs.getNamedItem('timestamp') &&
          attrs.getNamedItem('timestamp').value,
        user: attrs.getNamedItem('user') && attrs.getNamedItem('user').value,
        uid: attrs.getNamedItem('uid') && attrs.getNamedItem('uid').value
      }),
      tags: getTags(obj),
      members: getMembers(obj)
    });
  }
};
