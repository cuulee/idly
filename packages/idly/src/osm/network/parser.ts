import { osmNode, osmWay, osmRelation, osmEntity } from 'src/osm/entity';
function getLoc(attrs) {
  var lon = attrs.lon && attrs.lon.value,
    lat = attrs.lat && attrs.lat.value;
  return [parseFloat(lon), parseFloat(lat)];
}
function getNodes(obj) {
  var elems = obj.getElementsByTagName('nd'),
    nodes = new Array(elems.length);
  for (var i = 0, l = elems.length; i < l; i++) {
    nodes[i] = 'n' + elems[i].attributes.ref.value;
  }
  return nodes;
}

function getTags(obj) {
  var elems = obj.getElementsByTagName('tag'),
    tags = {};
  for (var i = 0, l = elems.length; i < l; i++) {
    var attrs = elems[i].attributes;
    tags[attrs.k.value] = attrs.v.value;
  }
  return tags;
}

function getMembers(obj) {
  var elems = obj.getElementsByTagName('member'),
    members = new Array(elems.length);
  for (var i = 0, l = elems.length; i < l; i++) {
    var attrs = elems[i].attributes;
    members[i] = {
      id: attrs.type.value[0] + attrs.ref.value,
      type: attrs.type.value,
      role: attrs.role.value
    };
  }
  return members;
}

function getVisible(attrs) {
  return !attrs.visible || attrs.visible.value !== 'false';
}

var parsers = {
  node: function nodeData(obj) {
    var attrs = obj.attributes;
    return new osmNode({
      id: osmEntity.id.fromOSM('node', attrs.id.value),
      visible: getVisible(attrs),
      version: attrs.version.value,
      changeset: attrs.changeset && attrs.changeset.value,
      timestamp: attrs.timestamp && attrs.timestamp.value,
      user: attrs.user && attrs.user.value,
      uid: attrs.uid && attrs.uid.value,
      loc: getLoc(attrs),
      tags: getTags(obj)
    });
  },

  way: function wayData(obj) {
    var attrs = obj.attributes;
    return new osmWay({
      id: osmEntity.id.fromOSM('way', attrs.id.value),
      visible: getVisible(attrs),
      version: attrs.version.value,
      changeset: attrs.changeset && attrs.changeset.value,
      timestamp: attrs.timestamp && attrs.timestamp.value,
      user: attrs.user && attrs.user.value,
      uid: attrs.uid && attrs.uid.value,
      tags: getTags(obj),
      nodes: getNodes(obj)
    });
  },

  relation: function relationData(obj) {
    var attrs = obj.attributes;
    return new osmRelation({
      id: osmEntity.id.fromOSM('relation', attrs.id.value),
      visible: getVisible(attrs),
      version: attrs.version.value,
      changeset: attrs.changeset && attrs.changeset.value,
      timestamp: attrs.timestamp && attrs.timestamp.value,
      user: attrs.user && attrs.user.value,
      uid: attrs.uid && attrs.uid.value,
      tags: getTags(obj),
      members: getMembers(obj)
    });
  }
};

export function parse(xml) {
  if (!xml || !xml.childNodes) return;

  var root = xml.childNodes[0],
    children = root.childNodes,
    entities = [];

  for (var i = 0, l = children.length; i < l; i++) {
    var child = children[i],
      parser = parsers[child.nodeName];
    if (parser) {
      entities.push(parser(child));
    }
  }

  return entities;
}
