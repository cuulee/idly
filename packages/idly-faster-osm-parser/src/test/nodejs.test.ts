import * as fs from 'fs';
import equal from 'lodash-es/isEqual';
import { iDParser } from '../iDParser';
import { nodejsParser } from '../nodejsParser';
import { parser } from '../parser';
import * as xmls from './fixtures';

function writeFile(path: string, data: string) {
  fs.writeFileSync(path, data, 'utf-8');
}

function deleteFile(path: string) {
  fs.unlinkSync(path);
}

describe('node and regular parity', () => {
  Object.keys(xmls)
    .map(k => [k, (xmls as any)[k]])
    .forEach(([k, xml]) => {
      it(`should work for ${k}`, () => {
        const path = `./xml.osm`;
        writeFile(path, xml);

        const nodeOutput = Array.from(nodejsParser(path));
        const regularOutput = parser(xml);
        const parsedXML = new DOMParser().parseFromString(
          xml
            .split('\n')
            .slice(1)
            .join('\n'),
          'text/xml'
        );
        const legacyOutput = iDParser(parsedXML);
        deleteFile(path);
        expect(equal(nodeOutput, regularOutput)).toBe(true);
        expect(equal(regularOutput, legacyOutput)).toBe(true);
        expect(nodeOutput.length).toMatchSnapshot();
      });
    });
});
