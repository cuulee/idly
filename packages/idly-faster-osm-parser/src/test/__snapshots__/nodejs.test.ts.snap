// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`node and regular parity should work for BIG_XML1 1`] = `1491`;

exports[`node and regular parity should work for BIG_XML2 1`] = `3070`;

exports[`node and regular parity should work for BIG_XML3 1`] = `3599`;

exports[`node and regular parity should work for NODE_XML1 1`] = `20`;

exports[`node and regular parity should work for WAY_XML1 1`] = `1`;

exports[`node and regular parity should work for XML1 1`] = `2`;

exports[`node and regular parity should work for XML2 1`] = `1`;

exports[`node and regular parity should work for XML3 1`] = `5`;

exports[`node and regular parity should work for XML5 1`] = `76`;

exports[`node and regular parity should work for XML7 1`] = `53`;

exports[`node and regular parity should work for XML8 1`] = `345`;

exports[`node and regular parity should work for XML9 1`] = `1936`;
